pallets/connectors/src/lib.rs

```diff
@@ -612,11 +865,108 @@ pub mod pallet {
+		///
+		/// Assumes the currency to be registered in the `AssetRegistry`.
+		///
+		/// NOTE: Reverse operation of [try_get_general_index].
+		pub fn try_get_currency_id(
+			index: GeneralCurrencyIndexOf<T>,
+		) -> Result<CurrencyIdOf<T>, DispatchError> {
+			let currency = CurrencyIdOf::<T>::try_from(index)?;
+			ensure!(
+				T::AssetRegistry::metadata(&currency).is_some(),
+				Error::<T>::AssetNotFound
+			);
+
+			Ok(currency)
+		}
+
+		pub fn try_get_wrapped_currency(
+			currency_id: &CurrencyIdOf<T>,
+		) -> Result<ConnectorsWrappedCurrency, DispatchError> {
+			let meta = T::AssetRegistry::metadata(&currency_id).ok_or(Error::<T>::AssetNotFound)?;
+			ensure!(
+				meta.additional.transferability.includes_connectors(),
+				Error::<T>::AssetNotConnectorsTransferable
+			);
+			T::CurrencyConverter::convert(meta.location.ok_or(Error::<T>::InvalidTransferCurrency)?)
+				.map_err(|_| Error::<T>::AssetNotConnectorsWrappedCurrency.into())
+		}
+
+		/// Ensures that the given pool and tranche exists and returns the
+		/// corresponding investment id.
+		pub fn derive_invest_id(
+			pool_id: PoolIdOf<T>,
+			tranche_id: TrancheIdOf<T>,
+		) -> Result<<T as pallet::Config>::TrancheCurrency, DispatchError> {
+			ensure!(
+				T::PoolInspect::pool_exists(pool_id),
+				Error::<T>::PoolNotFound
+			);
+			ensure!(
+				T::PoolInspect::tranche_exists(pool_id, tranche_id),
+				Error::<T>::TrancheNotFound
+			);
+
+			Ok(TrancheCurrency::generate(pool_id, tranche_id))
+		}
+
+		/// Ensures that the payment currency of the given investment id matches
+		/// the derived currency and returns the latter.
+		pub fn try_get_payment_currency(
+			invest_id: <T as pallet::Config>::TrancheCurrency,
+			currency_index: GeneralCurrencyIndexOf<T>,
+		) -> Result<CurrencyIdOf<T>, DispatchError> {
+			// retrieve currency id from general index
+			let currency = Self::try_get_currency_id(currency_index)?;
+
+			// get investment info
+			let payment_currency: CurrencyIdOf<T> =
+				<T as pallet::Config>::ForeignInvestmentAccountant::info(invest_id)?
+					.payment_currency()
+					.into();
+			ensure!(
+				payment_currency == currency,
+				Error::<T>::InvalidInvestCurrency
+			);
+
+			Ok(currency)
+		}
+	}
+
+	impl<T: Config> cfg_traits::ForeignInvestments<CurrencyIdOf<T>> for Pallet<T> {
+		type AssetRegistry = T::AssetRegistry;
+		type Error = DispatchError;
+		type PoolId = PoolIdOf<T>;
+
+		fn can_invest_currency_into_pool(
+			pool_id: PoolIdOf<T>,
+			currency: CurrencyIdOf<T>,
+		) -> DispatchResult {
+			ensure!(
+				T::PoolInspect::pool_exists(pool_id),
+				Error::<T>::PoolNotFound
+			);
+
+			// Ensure metadata pool currency flag is enabled
+			let metadata =
+				T::AssetRegistry::metadata(&currency).ok_or(Error::<T>::AssetNotFound)?;
+			ensure!(
+				metadata.additional.pool_currency,
+				Error::<T>::AssetMetadataNotPoolCurrency
+			);
+
+			T::PoolInspect::currency_for(pool_id)
+				.filter(|c| c == &currency)
+				.map(|_| ())
```

Comment:
Do we need this `.map(|_| ())`?

Reply:
Got it 👍🏻

Reply:
Yeah I think it's cleaner than doing `filter_map` in this case. Of course, we could also do a simple `if else`.